KIPs/kip-71.md

```diff
@@ -0,0 +1,386 @@
+	address: int = 0
+	key: int = 0 
+	balance: int = 0
+	storage_root: int = 0
+	code_hash: int = 0
+	code_format: int = 0
+	vm_version: int = 0
+
+INITIAL_FORK_BLOCK_NUMBER = 107806544 # TBD
+BASE_FEE_DELTA_REDUCING_DENOMINATOR = 36 # Max basefee change: 5% per block 
+LOWER_BOUND_BASE_FEE = 25000000000 # 25ston
+UPPER_BOUND_BASE_FEE = 750000000000 # 750ston
+GAS_TARGET = 30000000 
+BLOCK_GAS_LIMIT = 84000000 # implicit block gas limit to enforce the max basefee change rate
+BURN_RATIO = 0.5 
+CN_DISTRIBUTION_RATIO = 0.34 # set by governance
+KGF_DISTRIBUTION_RATIO = 0.54 # set by governance
+KIR_DISTRIBUTION_RATIO = 0.12 # set by governance
+
+class World(ABC):
+	def validate_block(self, block: Block) -> None:
+
+		parent_base_fee_per_gas = self.parent(block).base_fee_per_gas
+		transactions = self.transactions(block)
+		parent_gas_used = self.parent(block).gas_used
+
+		if parent_gas_used > BLOCK_GAS_LIMIT:
+			parent_gas_used = BLOCK_GAS_LIMIT
+
+		# check if the base fee is correct
+		if INITIAL_FORK_BLOCK_NUMBER == block.number:
+			expected_base_fee_per_gas = LOWER_BOUND_BASE_FEE
+
+		elif parent_gas_used == GAS_TARGET:
+			expected_base_fee_per_gas = parent_base_fee_per_gas
+
+		elif parent_gas_used > GAS_TARGET:
+			gas_used_delta = parent_gas_used - GAS_TARGET
+			base_fee_per_gas_delta = max(parent_base_fee_per_gas * gas_used_delta // GAS_TARGET // BASE_FEE_DELTA_REDUCING_DENOMINATOR, 1)
+			expected_base_fee_per_gas = parent_base_fee_per_gas + base_fee_per_gas_delta
+			if expected_base_fee_per_gas > UPPER_BOUND_BASE_FEE:
+				expected_base_fee_per_gas = UPPER_BOUND_BASE_FEE
+			
+		else:
+			gas_used_delta = GAS_TARGET - parent_gas_used
+			base_fee_per_gas_delta = parent_base_fee_per_gas * gas_used_delta // GAS_TARGET // BASE_FEE_DELTA_REDUCING_DENOMINATOR
+			expected_base_fee_per_gas = parent_base_fee_per_gas - base_fee_per_gas_delta
+			if expected_base_fee_per_gas < LOWER_BOUND_BASE_FEE:
+				expected_base_fee_per_gas = LOWER_BOUND_BASE_FEE
+
+		assert expected_base_fee_per_gas == block.base_fee_per_gas, 'invalid block: base fee not correct'
+
+		# execute transactions and do gas accounting
+		cumulative_transaction_gas_used = 0
+		for unnormalized_transaction in transactions:
+			# Note: this validates transaction signature and chain ID which must happen before we normalize below since normalized transactions don't include signature or chain ID
+			signer_address = self.validate_and_recover_signer_address(unnormalized_transaction)
+			transaction = self.normalize_transaction(unnormalized_transaction, signer_address)
+
+			signer = self.account(signer_address)
+
+			
+			signer.balance -= transaction.amount
+			assert signer.balance >= 0, 'invalid transaction: signer does not have enough ETH to cover attached value'
+			
+			# TODO: fee delegation transaction accounting
+			# the signer must be able to afford the transaction
+			assert signer.balance >= transaction.gas_limit * transaction.max_fee_per_gas
```

Comment:
```suggestion
			assert fee_payer.balance >= transaction.gas_limit * transaction.max_fee_per_gas
```

we can replace signer to fee_payer so that we can include fee delegation transactions. i.e., we can think of a sender of non-fee-delegated transactions as a fee-payer.